File: bindings/zig/src/commands.zig
const std = @import("std");

const uhppote = @import("uhppote/uhppote.zig");
const decode = @import("uhppote/decode.zig");
const datelib = @import("uhppote/datetime.zig");
const network = @import("uhppote/network.zig");

const CONTROLLER: u32 = 405419896;
const DOOR: u8 = 3;
const MODE: u8 = 2;
const DELAY: u8 = 10;
const CARD: u32 = 8165538;
const CARD_INDEX: u32 = 3;
const EVENT_INDEX: u32 = 37;
const TIME_PROFILE_ID: u8 = 29;

var CONTROLLERS = std.AutoHashMapUnmanaged(u32, uhppote.Controller){};

pub const Command = struct {
    name: []const u8,
    function: *const fn (std.mem.Allocator) void,
};

pub const commands = [_]Command{
    Command{
        .name = "get-all-controllers",
        .function = get_all_controllers,
    },

    Command{
        .name = "get-controller",
        .function = get_controller,
    },

    Command{
        .name = "set-ip",
        .function = set_ip,
    },

    Command{
        .name = "get-time",
        .function = get_time,
    },

    Command{
        .name = "set-time",
        .function = set_time,
    },

    Command{
        .name = "get-status",
        .function = get_status,
    },

    Command{
        .name = "get-listener",
        .function = get_listener,
    },

    Command{
        .name = "set-listener",
        .function = set_listener,
    },

    Command{
        .name = "get-door-control",
        .function = get_door_control,
    },

    Command{
        .name = "set-door-control",
        .function = set_door_control,
    },

    Command{
        .name = "open-door",
        .function = open_door,
    },

    Command{
        .name = "get-cards",
        .function = get_cards,
    },

    Command{
        .name = "get-card",
        .function = get_card,
    },

    Command{
        .name = "get-card-by-index",
        .function = get_card_by_index,
    },

    Command{
        .name = "put-card",
        .function = put_card,
    },

    Command{
        .name = "delete-card",
        .function = delete_card,
    },

    Command{
        .name = "delete-all-cards",
        .function = delete_all_cards,
    },

    Command{
        .name = "get-event",
        .function = get_event,
    },

    Command{
        .name = "get-event-index",
        .function = get_event_index,
    },

    Command{
        .name = "set-event-index",
        .function = set_event_index,
    },

    Command{
        .name = "record-special-events",
        .function = record_special_events,
    },

    Command{
        .name = "get-time-profile",
        .function = get_time_profile,
    },

    Command{
        .name = "set-time-profile",
        .function = set_time_profile,
    },

    Command{
        .name = "delete-all-time-profiles",
        .function = delete_all_time_profiles,
    },

    Command{
        .name = "add-task",
        .function = add_task,
    },

    Command{
        .name = "refresh-tasklist",
        .function = refresh_tasklist,
    },

    Command{
        .name = "clear-tasklist",
        .function = clear_tasklist,
    },

    Command{
        .name = "set-pc-control",
        .function = set_pc_control,
    },

    Command{
        .name = "set-interlock",
        .function = set_interlock,
    },

    Command{
        .name = "activate-keypads",
        .function = activate_keypads,
    },

    Command{
        .name = "set-door-passcodes",
        .function = set_door_passcodes,
    },

    Command{
        .name = "restore-default-parameters",
        .function = restore_default_parameters,
    },

    Command{
        .name = "listen",
        .function = listen,
    },
};

pub fn init(allocator: std.mem.Allocator) !void {
    try CONTROLLERS.put(allocator,405419896, uhppote.Controller {
        .controller = 405419896,
        .address = "192.168.1.100:60000",
        .transport = "tcp",
    });

    try CONTROLLERS.put(allocator,303986753, uhppote.Controller {
        .controller = 303986753,
        .address = "192.168.1.100:60000",
        .transport = "udp",
    });
}

pub fn deinit(allocator: std.mem.Allocator) void {
    CONTROLLERS.deinit(allocator);
}

pub fn exec(cmd: Command) !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();

    const allocator = gpa.allocator();

    cmd.function(allocator);
}

fn get_all_controllers(allocator: std.mem.Allocator) void {
    const list = uhppote.get_all_controllers(allocator);
    if (list) |l| {
        for (l) |controller| {
            pprint(controller);
        }

        allocator.free(l);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_controller(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.get_controller(controller, allocator)) |c| {
        pprint(c);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_ip(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const address = network.Address.IPv4.init(192, 168, 1, 100);
    const netmask = network.Address.IPv4.init(255, 255, 255, 0);
    const gateway = network.Address.IPv4.init(192, 168, 1, 1);

    if (uhppote.set_ip(controller, address, netmask, gateway, allocator)) |ok| {
        pprint(ok);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_time(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.get_time(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_time(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const now = datelib.now();

    if (uhppote.set_time(controller, now, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_status(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.get_status(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_listener(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.get_listener(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_listener(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const address = network.Address.IPv4.init(192, 168, 1, 100);
    const port = 60002;
    const interval = 15;

    if (uhppote.set_listener(controller, address, port, interval, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_door_control(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const door = DOOR;

    if (uhppote.get_door_control(controller, door, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_door_control(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const door = DOOR;
    const mode = MODE;
    const delay = DELAY;

    if (uhppote.set_door_control(controller, door, mode, delay, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn open_door(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const door = DOOR;

    if (uhppote.open_door(controller, door, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_cards(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.get_cards(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_card(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const card = CARD;

    if (uhppote.get_card(controller, card, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_card_by_index(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const index = CARD_INDEX;

    if (uhppote.get_card_by_index(controller, index, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn put_card(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const card = CARD;
    const start = datelib.Date{ .year = 2023, .month = 1, .day = 1 };
    const end = datelib.Date{ .year = 2023, .month = 12, .day = 31 };
    const door1 = 0;
    const door2 = 1;
    const door3 = 29;
    const door4 = 0;
    const pin = 7531;

    if (uhppote.put_card(controller, card, start, end, door1, door2, door3, door4, pin, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn delete_card(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const card = CARD;

    if (uhppote.delete_card(controller, card, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn delete_all_cards(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.delete_all_cards(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_event(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const index = EVENT_INDEX;

    if (uhppote.get_event(controller, index, allocator)) |response| {
        if (response.event_type == 0xff) {
            std.debug.print("\n   *** ERROR  event @ index {any} overwritten\n", .{index});
        } else if (response.index == 0) {
            std.debug.print("\n   *** ERROR  event @ index {any} not found\n", .{index});
        } else {
            pprint(response);
        }
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_event_index(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.get_event_index(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_event_index(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const index = EVENT_INDEX;

    if (uhppote.set_event_index(controller, index, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn record_special_events(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const enabled = true;

    if (uhppote.record_special_events(controller, enabled, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn get_time_profile(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const profile_id = TIME_PROFILE_ID;

    if (uhppote.get_time_profile(controller, profile_id, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_time_profile(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const profile_id = TIME_PROFILE_ID;
    const start = datelib.Date{ .year = 2023, .month = 1, .day = 1 };
    const end = datelib.Date{ .year = 2023, .month = 12, .day = 31 };
    const monday = true;
    const tuesday = true;
    const wednesday = false;
    const thursday = true;
    const friday = false;
    const saturday = false;
    const sunday = true;
    const segment_1_start = datelib.Time{ .hour = 8, .minute = 30, .second = 0 };
    const segment_1_end = datelib.Time{ .hour = 11, .minute = 45, .second = 0 };
    const segment_2_start = datelib.Time{ .hour = 12, .minute = 45, .second = 0 };
    const segment_2_end = datelib.Time{ .hour = 17, .minute = 0, .second = 0 };
    const segment_3_start = datelib.Time{ .hour = 18, .minute = 30, .second = 0 };
    const segment_3_end = datelib.Time{ .hour = 20, .minute = 15, .second = 0 };
    const linked_profile_id = 37;

    if (uhppote.set_time_profile(controller, profile_id, start, end, monday, tuesday, wednesday, thursday, friday, saturday, sunday, segment_1_start, segment_1_end, segment_2_start, segment_2_end, segment_3_start, segment_3_end, linked_profile_id, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn delete_all_time_profiles(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.delete_all_time_profiles(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn add_task(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const start_date = datelib.Date{ .year = 2023, .month = 1, .day = 1 };
    const end_date = datelib.Date{ .year = 2023, .month = 12, .day = 31 };
    const monday = true;
    const tuesday = true;
    const wednesday = false;
    const thursday = true;
    const friday = false;
    const saturday = false;
    const sunday = true;
    const start_time = datelib.Time{ .hour = 8, .minute = 30, .second = 0 };
    const door = DOOR;
    const task_type = 2;
    const more_cards = 0;

    if (uhppote.add_task(controller, start_date, end_date, monday, tuesday, wednesday, thursday, friday, saturday, sunday, start_time, door, task_type, more_cards, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn refresh_tasklist(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.refresh_tasklist(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn clear_tasklist(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.clear_tasklist(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_pc_control(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const enabled = true;

    if (uhppote.set_pc_control(controller, enabled, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_interlock(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const interlock = 3;

    if (uhppote.set_interlock(controller, interlock, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn activate_keypads(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const reader1 = true;
    const reader2 = true;
    const reader3 = false;
    const reader4 = true;

    if (uhppote.activate_keypads(controller, reader1, reader2, reader3, reader4, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn set_door_passcodes(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);
    const door = DOOR;
    const passcode1 = 12345;
    const passcode2 = 0;
    const passcode3 = 999999;
    const passcode4 = 54321;

    if (uhppote.set_door_passcodes(controller, door, passcode1, passcode2, passcode3, passcode4, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn restore_default_parameters(allocator: std.mem.Allocator) void {
    const controller = resolve(CONTROLLER);

    if (uhppote.restore_default_parameters(controller, allocator)) |response| {
        pprint(response);
    } else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn listen(allocator: std.mem.Allocator) void {
    if (uhppote.listen(on_event, allocator)) {} else |err| {
        std.debug.print("\n   *** ERROR  {any}\n", .{err});
    }
}

fn on_event(event: decode.Event) void {
    pprint(event);
}

fn resolve(controller: u32) uhppote.Controller {
    return CONTROLLERS.get(controller) orelse uhppote.Controller {
        .controller = controller,
        .address = "",
        .transport = "udp",
    };
}

fn pprint(v: anytype) void {
    std.debug.print("{any}\n", .{v});
}

File: bindings/go/commands.go
package main

import (
"strconv"
    "encoding/json"
    "fmt"
    "log"
    "net/netip"
    "os"
    "os/signal"
    "syscall"
    "time"

    "uhppoted/uhppote"
)

const CONTROLLER uint32 = 405419896
const DOOR uint8 = 3
const MODE uint8 = 2
const DELAY uint8 = 10
const CARD uint32 = 8165538
const CARD_INDEX uint32 = 3
const EVENT_INDEX uint32 = 37
const TIME_PROFILE_ID uint8 = 29

var ADDRESS = netip.MustParseAddr("192.168.1.100")
var NETMASK = netip.MustParseAddr("255.255.255.0")
var GATEWAY = netip.MustParseAddr("192.168.1.1")
var LISTENER = netip.MustParseAddrPort("192.168.1.100:60001")

var controllers = map[uint32]uhppote.Controller{
    405419896: uhppote.Controller{405419896, "192.168.1.100:60000", "tcp"},
}

type command struct {
    name string
    f    func(args []string) (any, error)
}

var commands = []command{
    command{name: "get-all-controllers", f: getAllControllers},
    command{name: "get-controller", f: getController},
    command{name: "set-ip", f: setIP},
    command{name: "get-time", f: getTime},
    command{name: "set-time", f: setTime},
    command{name: "get-listener", f: getListener},
    command{name: "set-listener", f: setListener},
    command{name: "get-door-control", f: getDoorControl},
    command{name: "set-door-control", f: setDoorControl},
    command{name: "get-status", f: getStatus},
    command{name: "open-door", f: openDoor},
    command{name: "get-cards", f: getCards},
    command{name: "get-card", f: getCard},
    command{name: "get-card-by-index", f: getCardByIndex},
    command{name: "put-card", f: putCard},
    command{name: "delete-card", f: deleteCard},
    command{name: "delete-all-cards", f: deleteAllCards},
    command{name: "get-event", f: getEvent},
    command{name: "get-event-index", f: getEventIndex},
    command{name: "set-event-index", f: setEventIndex},
    command{name: "record-special-events", f: recordSpecialEvents},
    command{name: "get-time-profile", f: getTimeProfile},
    command{name: "set-time-profile", f: setTimeProfile},
    command{name: "delete-all-time-profiles", f: deleteAllTimeProfiles},
    command{name: "add-task", f: addTask},
    command{name: "refresh-tasklist", f: refreshTaskList},
    command{name: "clear-tasklist", f: clearTaskList},
    command{name: "set-pc-control", f: setPCControl},
    command{name: "set-interlock", f: setInterlock},
    command{name: "activate-keypads", f: activateKeypads},
    command{name: "set-door-passcodes", f: setDoorPasscodes},
    command{name: "restore-default-parameters", f: restoreDefaultParameters},
    command{name: "listen", f: listen},
}

func (c command) exec(args []string) {
    if response, err := c.f(args); err != nil {
        log.Fatalf("ERROR  %v", err)
    } else if response != nil {
        log.Printf("INFO  %+v", pprint(response))
    } else {
        log.Printf("WARN  no response")
    }
}

func pprint(v any) string {
    if bytes, err := json.MarshalIndent(v, "", "  "); err != nil {
        return fmt.Sprintf("%v", v)
    } else {
        return string(bytes)
    }
}

func parseArgs(args []string, arg string, defval any) any {
    ix := 0
    for ix < len(args) {
        if args[ix] == arg {
            ix++
            if ix < len(args) {
                switch defv := defval.(type) {
                case uint32:
                    if v,err := strconv.ParseUint(args[ix],10,32); err != nil {
                        return defv
                    } else {
                        return uint32(v)
                    }

                default:
                    return defval
                }
            }
        }
       ix++
    }

    return defval
}

func getAllControllers(args []string) (any, error) {
    return uhppote.GetAllControllers()
}

func getController(args []string) (any, error) {
    controller := resolve(parseArgs(args,"--controller", CONTROLLER).(uint32))

    return uhppote.GetController(controller)
}

func setIP(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    address := ADDRESS
    netmask := NETMASK
    gateway := GATEWAY

    if err := uhppote.SetIP(controller, address, netmask, gateway); err != nil {
        return nil, err
    }

    return nil, nil
}

func getTime(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.GetTime(controller)
}

func setTime(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    now := uhppote.DateTime(time.Now())

    return uhppote.SetTime(controller, now)
}

func getStatus(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.GetStatus(controller)
}

func getListener(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.GetListener(controller)
}

func setListener(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    address := LISTENER.Addr()
    port := LISTENER.Port()

    return uhppote.SetListener(controller, address, port)
}

func getDoorControl(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    door := DOOR

    return uhppote.GetDoorControl(controller, door)
}

func setDoorControl(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    door := DOOR
    mode := MODE
    delay := DELAY

    return uhppote.SetDoorControl(controller, door, mode, delay)
}

func openDoor(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    door := DOOR

    return uhppote.OpenDoor(controller, door)
}

func getCards(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.GetCards(controller)
}

func getCard(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    card := CARD

    if response,err := uhppote.GetCard(controller, card); err != nil {
        return nil, err
    } else if response.CardNumber == 0 {
        return nil, fmt.Errorf("card %v not found", card)
    } else {
        return response, nil
    }
}

func getCardByIndex(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    index := CARD_INDEX

    if response,err := uhppote.GetCardByIndex(controller, index); err != nil {
        return nil, err
    } else if response.CardNumber == 0 {
        return nil, fmt.Errorf("card @ index %v not found", index)
    } else if response.CardNumber == 0xffffffff {
        return nil, fmt.Errorf("card @ index %v deleted", index)
    } else {
        return response, nil
    }
}

func putCard(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    card := CARD
    start, _ := time.Parse("2006-01-02", "2022-01-01")
    end, _ := time.Parse("2006-01-02", "2022-12-31")
    pin := uhppote.PIN(7531)

    return uhppote.PutCard(controller, card, uhppote.Date(start), uhppote.Date(end), 0, 1, 29, 0,pin)
}

func deleteCard(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    card := CARD

    return uhppote.DeleteCard(controller, card)
}

func deleteAllCards(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.DeleteAllCards(controller)
}

func getEvent(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    index := EVENT_INDEX

    if response, err := uhppote.GetEvent(controller, index); err != nil {
        return nil, err
    } else if response.EventType == 0xff {
        return nil, fmt.Errorf("event @ index %v overwritten", index)
    } else if response.Index == 0 {
        return nil, fmt.Errorf("event @ index %v not found", index)
    } else {
        return response, nil
    }
}

func getEventIndex(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.GetEventIndex(controller)
}

func setEventIndex(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    index := EVENT_INDEX

    return uhppote.SetEventIndex(controller, index)
}

func recordSpecialEvents(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    enabled := true

    return uhppote.RecordSpecialEvents(controller, enabled)
}

func getTimeProfile(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    profileID := TIME_PROFILE_ID

    if response,err := uhppote.GetTimeProfile(controller, profileID); err != nil {
        return nil, err
    } else if response.ProfileId == 0 {
        return nil, fmt.Errorf("time profile %v not defined", profileID)
    } else {
        return response, nil
    }
}

func setTimeProfile(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    profileID := TIME_PROFILE_ID
    start, _ := time.Parse("2006-01-02", "2022-01-01")
    end, _ := time.Parse("2006-01-02", "2022-12-31")
    monday := true
    tuesday := true
    wednesday := false
    thursday := true
    friday := false
    saturday := false
    sunday := true
    segment1start, _ := time.Parse("15:04", "08:30")
    segment1end, _ := time.Parse("15:04", "11:45")
    segment2start, _ := time.Parse("15:04", "13:15")
    segment2end, _ := time.Parse("15:04", "16:30")
    segment3start, _ := time.Parse("15:04", "19:30")
    segment3end, _ := time.Parse("15:04", "20:55")
    linkedProfileID := uint8(30)

    return uhppote.SetTimeProfile(
        controller,
        profileID,
        uhppote.Date(start), uhppote.Date(end),
        monday, tuesday, wednesday, thursday, friday, saturday, sunday,
        uhppote.HHmm(segment1start), uhppote.HHmm(segment1end),
        uhppote.HHmm(segment2start), uhppote.HHmm(segment2end),
        uhppote.HHmm(segment3start), uhppote.HHmm(segment3end),
        linkedProfileID)
}

func deleteAllTimeProfiles(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.DeleteAllTimeProfiles(controller)
}

func addTask(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    startDate, _ := time.Parse("2006-01-02", "2022-01-01")
    endDate, _ := time.Parse("2006-01-02", "2022-12-31")
    monday := true
    tuesday := false
    wednesday := true
    thursday := true
    friday := false
    saturday := false
    sunday := true
    startTime, _ := time.Parse("15:04", "08:30")
    door := DOOR
    taskType := uint8(2)
    moreCards := uint8(0)

    return uhppote.AddTask(controller,
        uhppote.Date(startDate), uhppote.Date(endDate),
        monday, tuesday, wednesday, thursday, friday, saturday, sunday,
        uhppote.HHmm(startTime),
        door,
        taskType,
        moreCards)
}

func refreshTaskList(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.RefreshTasklist(controller)
}

func clearTaskList(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.ClearTasklist(controller)
}

func setPCControl(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    enabled := true

    return uhppote.SetPcControl(controller, enabled)
}

func setInterlock(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    interlock := uint8(3)

    return uhppote.SetInterlock(controller, interlock)
}

func activateKeypads(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    reader1 := true
    reader2 := true
    reader3 := false
    reader4 := true

    return uhppote.ActivateKeypads(controller, reader1, reader2, reader3, reader4)
}

func setDoorPasscodes(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)
    door := DOOR
    passcode1 := uint32(12345)
    passcode2 := uint32(0)
    passcode3 := uint32(999999)
    passcode4 := uint32(54321)

    return uhppote.SetDoorPasscodes(controller, door, passcode1, passcode2, passcode3, passcode4)
}

func restoreDefaultParameters(args []string) (any, error) {
    controller := parseArgs(args,"--controller", CONTROLLER).(uint32)

    return uhppote.RestoreDefaultParameters(controller)
}

func listen(args []string) (any, error) {
    events := make(chan uhppote.Event)
    errors := make(chan error)
    interrupt := make(chan os.Signal, 1)

    defer close(events)
    defer close(errors)
    defer close(interrupt)

    signal.Notify(interrupt, syscall.SIGINT, syscall.SIGTERM)

    go func() {
        for evt := range events {
            log.Printf("INFO  %+v", pprint(evt))
        }
    }()

    go func() {
        for err := range errors {
            log.Fatalf("ERROR  %v", err)
        }
    }()

    if err := uhppote.Listen(events, errors, interrupt); err != nil {
        return nil, err
    }

    return struct{}{}, nil
}

func resolve(controller any) uhppote.Controller {
    switch v := controller.(type) {
    case uint32:
        if c, ok := controllers[v]; ok {
            return c
        } else {
            return uhppote.Controller{
                Controller: v,
                Address:    "",
                Transport:  "udp",
            }
        }

    case uhppote.Controller:
        return v
    }

    panic(fmt.Sprintf("unknown controller type (%T)", controller))
}
